lib/features/build_server/cache/dependencies/dependency_graph.dart

```diff
@@ -0,0 +1,179 @@
+    final srcNode = _getNode(source);
+
+    for (final node in nodes) {
+      // add dependencies to source node
+      if (node == srcNode) {
+        node.dependencies = dependencies.toSet();
+        continue;
+      }
+
+      // add source node to it's dependencies or remove if no longer included
+      if (dependencies.contains(node.id)) {
+        node.dependents.add(srcNode.id);
+      } else {
+        node.dependents.remove(srcNode.id);
+      }
+    }
+  }
+
+  /// Returns a set of all nodes that depend on the node with the id [file].
+  Set<String> dependents(String file) => _getNode(file).dependents.toSet();
+
+  /// Returns a set of all nodes that depend even indirectly on the node with the id [file].
+  ///
+  /// Unlike `dependents`, also node ids are returned which are not directly dependent on [file].
+  /// If [file] is dependent on node `a` and node `a` is dependent on node `b`,
+  /// node `a` is returned as well as node `b`.
+  Set<String> indirectDependents(String file) {
+    final result = <String>{};
+
+    void addDepentents(_DependencyNode node) {
+      for (final dependent in node.dependents.map((e) => _getNode(e))) {
+        if (!result.contains(dependent.id)) {
+          result.add(dependent.id);
+          addDepentents(dependent);
+        }
+      }
+    }
+
+    addDepentents(_getNode(file));
+
+    return result;
+  }
+
+  /// Returns a set of all nodes that [file] depends on.
+  Set<String> dependencies(String file) => _getNode(file).dependencies.toSet();
+
+  /// Returns a set of all nodes that [file] depends on even indirectly.
+  ///
+  /// If [file] depents on node `a` and node `a` is dependent on node `b`,
+  /// node `a` is returned as well as node `b`.
+  Set<String> indirectDependencies(String file) {
+    final result = <String>{};
+
+    void addDependencies(_DependencyNode node) {
+      for (final dependency in node.dependencies.map((e) => _getNode(e))) {
+        if (!result.contains(dependency.id)) {
+          result.add(dependency.id);
+          addDependencies(dependency);
+        }
+      }
+    }
+
+    addDependencies(_getNode(file));
+
+    return result;
+  }
+
+  /// Finds the node with the given [id] and returns it.
+  ///
+  /// If no node with the given id exists, an error is thrown.
+  _DependencyNode _getNode(String id) {
+    for (final node in nodes) {
+      if (node.id == id) {
+        return node;
+      }
+    }
+
+    throw ArgumentError('No node with id "$id" exists');
+  }
+
+  /// Converts a set of ids to a set of nodes without any dependencies.
+  static Set<_DependencyNode> _convert(Set<String> ids) =>
+      ids.map((id) => _DependencyNode(id: id, dependencies: {}, dependents: {})).toSet();
+}
+
+/// A single node representing a source file with
+/// different dependencies and dependents.
+@JsonSerializable()
+class _DependencyNode {
+  /// The id of the node.
+  ///
+  /// Sole parameter for equality checks.
+  final String id;
```

Comment:
Actually it's really just the id of the node and `_DependencyNode` is not the instance in the code which determines what to use as the id. That's the job of `DependencyGraph`. Finally it will be set to the 'id' of each `SourceFile`.

Would you prefer me to add an implementation hint aka `It is recommended to set this field to the unique id of the corresponding source file` or similar?

Reply:
Yes, such a hint would be great because it seems a bit arbitrary atm.